pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go

```diff
@@ -210,19 +215,97 @@ func (state PipelineRunState) GetNextTasks(candidateTasks map[string]struct{}) [
 	return tasks
 }
 
-// SuccessfulPipelineTaskNames returns a list of the names of all of the PipelineTasks in state
-// which have successfully completed.
-func (state PipelineRunState) SuccessfulPipelineTaskNames() []string {
-	done := []string{}
+// SuccessfulOrSkippedDAGTasks returns a list of the names of all of the PipelineTasks in state
+// which have successfully completed or skipped
+func (state PipelineRunState) SuccessfulOrSkippedDAGTasks(d *dag.Graph) []string {
+	tasks := []string{}
 	for _, t := range state {
-		if t.TaskRun != nil {
-			c := t.TaskRun.Status.GetCondition(apis.ConditionSucceeded)
-			if c.IsTrue() {
-				done = append(done, t.PipelineTask.Name)
+		if _, ok := d.Nodes[t.PipelineTask.Name]; ok {
+			if t.IsSuccessful() || isSkipped(t, state.ToMap(), d) {
+				tasks = append(tasks, t.PipelineTask.Name)
+			}
+		}
+	}
+	return tasks
+}
+
+// isDAGTasksStopped returns true if any of the pipeline task has failed
+// and none of the pipleine task are still running
+func (state PipelineRunState) isDAGTasksStopped(d *dag.Graph) bool {
+	failed := false
+	for _, t := range state {
+		if t.IsFailure() {
+			failed = true
+			continue
+		}
+		if t.IsStarted() && !t.IsDone() {
+			failed = false
+			break
+		}
+	}
+	return failed
+}
+
+// isDAGTasksDone returns true if all DAG tasks are done executing,
+// DAG tasks could have failed/succeeded/skipped
+func (state PipelineRunState) isDAGTasksDone(d *dag.Graph) (isDone bool) {
+	isDone = true
+	for _, t := range state {
+		if _, ok := d.Nodes[t.PipelineTask.Name]; ok {
+			if t.TaskRun == nil {
+				// this task might have skipped if taskRun is nil
+				// continue and ignore if this task was skipped
+				// skipped task is considered part of done
+				if isSkipped(t, state.ToMap(), d) {
+					continue
+				}
+				return false
+			}
+			isDone = isDone && t.IsDone()
+			if !isDone {
+				return
+			}
+		}
+	}
+	return
+}
+
+// isFinalTasksDone returns true if all final tasks are done executing,
+// final tasks could have failed/succeeded
+func (state PipelineRunState) isFinalTasksDone(d *dag.Graph) (isDone bool) {
+	isDone = true
+	for _, t := range state {
+		if _, ok := d.Nodes[t.PipelineTask.Name]; ok {
+			if t.TaskRun == nil {
+				return false
+			}
+			isDone = isDone && t.IsDone()
+			if !isDone {
+				return
 			}
 		}
 	}
-	return done
+	return
+}
+
+// GetFinalTasks returns a list of final tasks without any taskRun associated with it
+// GetFinalTasks returns final tasks only when all DAG tasks have finished executing successfully or skipped or
+// any one DAG task resulted in failure
+func (state PipelineRunState) GetFinalTasks(d *dag.Graph, dfinally *dag.Graph) []*ResolvedPipelineRunTask {
+	tasks := []*ResolvedPipelineRunTask{}
+	finalCandidates := map[string]struct{}{}
+	// check either pipeline has finished executing all DAG pipelineTasks
+	// or any one of the DAG pipelineTask has failed
+	if state.isDAGTasksStopped(d) || state.isDAGTasksDone(d) {
```

Comment:
These two functions are only used here, they both loop over the pipelineState and both have the same exact signature. Is there any reason to keep them separated?

Reply:
Both are serving different purposes:

`isDAGTasksStopped` checks if any of the pipeline task has failed and if so, no other pipeline tasks are still running

`isDAGTasksDone` checks if all DAG tasks are visited and done executing i.e. DAG tasks have succeeded, skipped or failed

Reply:
It seems to me that `isDAGTasksStopped` implies `isDAGTasksDone`, so I think `state.isDagDone(d)` would better represent the combination of the function. It's a function that checks if the subset of the state represented in the input DAG has completed its work.

From a naming POV, until now we always talked about `Tasks`, which are independent tasks, `PipelineTasks` which are tasks within a pipeline and `Nodes`, which are nodes in a DAG, realised as tasks. `DAGTasks` seem to mix the terminology across DAG and pipeline.

Reply:
i agree with @afrittoli about combining these into one function - i think both he and i stumbled a bit trying to understand the difference between the two functions, and it appears they are both called only in this one place, so it might be easier to understand overall to combine them both into something like `isDAGDone`

i actually think you could combine `isFinalTasksDone`, `isDAGTasksStopped` and `isDAGTasksDone` into one function, which i think you are suggesting above with `checkTasksDone`! :D

> DAGTasks seem to mix the terminology across DAG and pipeline.

@afrittoli i also feel a bit :( about this but it seems like the best option we've got - we've now got tasks that are executed as part of the pipeline graph, and tasks that are executed as part of finally. calling the later "finally tasks" seems to make sense, but for the former, just calling them "tasks" or "pipelinetasks" feels too ambiguous because that name applies to finally tasks too. dag tasks seems like the best option, but if you have another idea i think we'd definitely like something better (nonfinaltasks feels weird but is more descriptive)

^^ if we do decide to change the above, i suggest we do it in a follow up PR

Reply:
Tracking in issue #2852

Reply:
About the naming, "DAG Tasks" feels ambiguous to me too, because both the pipeline DAG and the finally tasks are represented by a dag. They're actually graph objects from the dag module, but we always refer to them as dags in parameters names and variables. Perhaps we could add a comment clearing the naming? It could be an improvement done on top.

About combining `isFinalTasksDone`, `isDAGTasksStopped` and `isDAGTasksDone`,  #2821 will remove the need of passing the dag, so we will end up with two functions `isFinalTasksDone` and isDAGTasksDone` that do not take any parameter.